filters/openpolicyagent/openpolicyagent.go

```diff
@@ -0,0 +1,393 @@
+package openpolicyagent
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+	"text/template"
+	"time"
+
+	ext_authz_v3_core "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/config"
+	"github.com/open-policy-agent/opa/logging"
+	"github.com/open-policy-agent/opa/plugins"
+	"github.com/open-policy-agent/opa/plugins/discovery"
+	"github.com/open-policy-agent/opa/rego"
+	"github.com/open-policy-agent/opa/runtime"
+	"github.com/open-policy-agent/opa/storage"
+	"github.com/open-policy-agent/opa/storage/inmem"
+	iCache "github.com/open-policy-agent/opa/topdown/cache"
+	opautil "github.com/open-policy-agent/opa/util"
+	"github.com/opentracing/opentracing-go"
+	"google.golang.org/protobuf/encoding/protojson"
+
+	"github.com/zalando/skipper/filters"
+	"github.com/zalando/skipper/filters/openpolicyagent/internal/envoy"
+	"github.com/zalando/skipper/filters/openpolicyagent/internal/util"
+)
+
+const (
+	OpenPolicyAgentDecisionKey        = "open-policy-agent:decision"
+	OpenPolicyAgentDecisionBodyKey    = "open-policy-agent:decision-body"
+	OpenPolicyAgentDecisionHeadersKey = "open-policy-agent:decision-headers"
+)
+
+type OpenPolicyAgentFactory struct {
+	// Ideally share one Bundle storage across many OPA "instances" using this factory.
```

Comment:
this can be done by using Registry style as in https://github.com/zalando/skipper/blob/master/ratelimit/registry.go or maybe even nicer in https://github.com/zalando/skipper/blob/master/scheduler/scheduler.go#L219 (fifo() / lifo() filters).

Another question is what should be done on route updates. If we have long running objects normally you do want to not drop the shared state.
If you want to handover filter instances from on route update to the next one lifo/fifo PostProcessor style can be used: 
filter https://github.com/zalando/skipper/blob/master/filters/scheduler/fifo.go
https://github.com/zalando/skipper/blob/master/scheduler/scheduler.go#L576